soc/xtensa/intel_adsp/common/include/cavs_hda.h

```diff
@@ -0,0 +1,303 @@
+/* Copyright (c) 2022 Intel Corporation
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_CAVS_HDA_H
+#define ZEPHYR_INCLUDE_CAVS_HDA_H
+
+#include <arch/xtensa/cache.h>
+#include <kernel.h>
+#include <device.h>
+#include <cavs-shim.h>
+#include <cavs-mem.h>
+
+/**
+ * @brief HDA stream functionality for cAVS
+ *
+ * Provides low level calls to support cAVS HDA streams with
+ * minimal abstraction that allows testing the hardware
+ * and its demands separately from the intended DMA API
+ * usage. Notably all addresses, counts, and sizes
+ * are predefined here if not defined elsewhere to enable
+ * easier unit testing.
+ */
+
+#ifndef HDA_HOST_OUT_BASE
+#define HDA_HOST_OUT_BASE 0x72800
+#endif
+
+#ifndef HDA_HOST_IN_BASE
+#define HDA_HOST_IN_BASE 0x72c00
+#endif
+
+#ifndef HDA_STREAM_COUNT
+#define HDA_STREAM_COUNT 7
+#endif
+
+#ifndef HDA_REGBLOCK_SIZE
+#define HDA_REGBLOCK_SIZE 0x40
+#endif
+
+/* The read/write positions are masked to 24 bits */
+#define HDA_RWP_MASK 0x00FFFFFF
+
+/* Buffers must be 128 byte aligned, this mask enforces that */
+#define HDA_ALIGN_MASK 0xFFFFFF80
+
+
+#define HDA_ADDR(base, stream) (base + stream*HDA_REGBLOCK_SIZE)
+
+
+/* Gateway Control and Status Register */
+#define DGCS(base, stream) ((volatile uint32_t *)HDA_ADDR(base, stream))
+#define DGCS_SCS BIT(31) /* Sample container size */
+#define DGCS_GEN BIT(26) /* Gateway Enable */
+#define DGCS_L1ETP BIT(25) /* L1 Enter Prevent */
+#define DGCS_L1EXP BIT(25) /* L1 Exit Prevent */
+#define DGCS_FWCB BIT(23) /* Firmware Control Buffer */
+#define DGCS_GBUSY BIT(15) /* Gateway Busy */
+#define DGCS_TE BIT(14) /* Transfer Error */
+#define DGCS_BSC BIT(11) /* Buffer Segment Completion */
+#define DGCS_BOR BIT(10) /* Buffer Overrun */
+#define DGCS_BUR BIT(10) /* Buffer Underrun */
+#define DGCS_BF BIT(9) /* Buffer Full */
+#define DGCS_BNE BIT(8) /* Buffer Not Empty */
+#define DGCS_FIFORDY BIT(5) /* Enable FIFO */
+#define DGCS_BSCIE BIT(3) /* Buffer Segment Completion Interrupt Enable */
+
+/* Gateway Buffer Base Address */
+#define DGBBA(base, stream) ((volatile uint32_t *)(HDA_ADDR(base, stream) + 0x04))
```

Comment:
Style thing: the rest of cavs likes to use C structs to define register blocks (a pattern I personally love, but will admit I didn't introduce).  Code in the rest of Zephyr tends to use offsets like this but with the sys_read/write() APIs.  Adding another scheme here ("macros returning pointer") is maybe a little confusing?

Reply:
This isn't the only driver to use a macro for a register returning a volatile pointer to the register. To be honest mapping this to structs, in my opinion, won't help build anymore clarity into the code here.